[TIMOB-26003] Android: Provide API to support opt-out of Analytics at runtime

ypbnv opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-26003

**Description:**
Expose optOut property for Analytics.

The aps-analytics.jar in this PR is built from:https://github.com/appcelerator/aps_sdk/pull/317

ypbnv committed:
Expose optOut property for Analytics.

build commented:
<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.3.0 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10058-3PXZJREICBZBXBIKCICH55VQXNF7FTSPHBHQ4IDSY5FHRM5X3TLA
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10058-3PXZJREICBZBXBIKCICH55VQXNF7FTSPHBHQ4IDSY5FHRM5X3TLA/tests/Resources/ti.ui.maskedimage.addontest.js
  27:3  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  32:3  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  33:3  warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10058-3PXZJREICBZBXBIKCICH55VQXNF7FTSPHBHQ4IDSY5FHRM5X3TLA/tests/Resources/ti.ui.picker.addontest.js
  11:5  warning  'should' is assigned a value but never used  no-unused-vars

✖ 4 problems (0 errors, 4 warnings)


Running "checkFormat:ios" (checkFormat) task

Running "checkFormat:android" (checkFormat) task
Fatal error: Error: Formatting incorrect on "android/modules/analytics/src/java/ti/modules/titanium/analytics/AnalyticsModule.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='1844' length='1'>&#10;	</replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

ypbnv committed:
Fix formatting.

garymathews reviewed (changes requested):
Update documentation to include `Ti.Analytics.optedOut` ?

garymathews reviewed (changes requested):
I think we need to agree on the API specification for Ti.Analytics optOut because this is varying per implementation:
```
Ti.Analytics.optedOut
Ti.Analytics.setOptedOut(boolean)
Ti.Analytics.getOptedOut()
```

For consistency like our other Titanium APIs

For example, the Windows implementation uses `isOptOut`:
https://github.com/appcelerator/titanium_mobile_windows/pull/1238

NOTE: I agree with the naming scheme in your PR, but set to request changes until we figure this out

infosia commented:
For the naming scheme, @hansemannn recently suggested following. I'm fine with it and will go with it for Windows.

```
Ti.Analytics.optedOut = true|false;
Ti.Analytics.setOptedOut(true|false);
Ti.Analytics.isOptedOut(); 
```

ypbnv committed:
Update with the latest aps_analytics jar.

garymathews reviewed (approved):
CR: **PASS**

ypbnv committed:
Move appEnrollEvent sending after evaluation of app.js
Remove sending of geoEvent.
Ensure Analytics are initialized before app.js evaluation, but before the first event is sent.

ypbnv committed:
Add latest version of aps-analytics.

ypbnv commented:
@garymathews I made some changes. Would you be able go go over them?
I moved all the events from Analytics to be send after the evalutaion of app.js in order to allow opting out before that.

The aps-analytics.jar from the latest commit is built from this commit:
https://github.com/appcelerator/aps_sdk/pull/317/commits/a26b8d6f64d62ebde3521119760f45eec5435ac2

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/TiLaunchActivity.java:
```diff
@@ -186,8 +187,10 @@ protected void onNewIntent(Intent intent)
 	protected void windowCreated(Bundle savedInstanceState)
 	{
 		super.windowCreated(savedInstanceState);
+		TiPlatformHelper.getInstance().initialize();
```
For safety, `null` check on instances?

lokeshchdhry commented:
@ypbnv , Can you please resolve the conflicts.

Commit:
[TIMOB-26003] Remove redundant Geolocation doAnalytics()

Commit:
[TIMOB-26027] Update aps-analytics.jar

Commit:
[TIMOB-26027] Update AndroidManifest.xml

lokeshchdhry commented:
@ypbnv , @garymathews , When I try to run an app with the artifact with this fix my app just stays at the splash & I am not seeing any relevant logs too. Am I missing something here?
 Will try to get some of the logs & put them here.

ssjsamir commented:
@lokeshchdhry I also see the issue when trying to run an android app on windows using the build from this PR. (App just stays on splash screen)

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| ios.Titanium.UI.WebView | loading | 10.002 | file:///Users/build/Librar |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

lokeshchdhry commented:
@ssjsamir, Its being investigated.

Commit:
[TIMOB-26003] Initialize APSAnalytics()

lokeshchdhry commented:
@garymathews , After the above fix the app launches stays on the splash screen & after some time I get the `App isn't responding` message.
```
-- Start application log -----------------------------------------------------
[INFO] :   TiApplication: (main) [0,0] checkpoint, app created.
[INFO] :   TiApplication: (main) [90,90] Titanium 7.3.0 (2018/05/24 10:56 undefined)
[INFO] :   MultiDex: VM with version 2.1.0 has multidex support
[INFO] :   MultiDex: Installing application
[INFO] :   MultiDex: VM has multidex support, MultiDex support library is disabled.
[INFO] :   DatabaseHelper: No value in database for platform key: 'unique_machine_id' returning supplied default ''
[INFO] :   DatabaseHelper: No value in database for platform key: 'hardware_machine_id' returning supplied default ''
[DEBUG] :  APSAnalyticsModel: Creating Database appcAnalytics.db
[TRACE] :  updating tiapp metadata with Appcelerator Platform...
[ERROR] :  TiApplication: (main) [1575,1665] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [3,1668] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [36,1704] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [15,1719] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [4,1723] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [1,1724] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [4,1728] No valid root or current activity found for application instance
[WARN] :   TiAndroid: (main) [3,1731] Application instance no longer available. Unable to get current activity.
[ERROR] :  TiApplication: (main) [0,1731] No valid root or current activity found for application instance
[WARN] :   TiAndroid: (main) [1,1732] Application instance no longer available. Unable to get current activity.
[WARN] :   TiAndroid: (main) [0,1732] Application instance no longer available. Unable to get current activity.
[ERROR] :  TiApplication: (main) [2,1734] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [4,1738] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [2,1740] No valid root or current activity found for application instance
[ERROR] :  TiApplication: (main) [2,1742] No valid root or current activity found for application instance
[INFO] :   TiApplication: (main) [16,1758] Titanium Javascript runtime: v8
[DEBUG] :  TouchtestmoduleModule: (main) [22,1780] inside onAppCreate
[DEBUG] :  CrittercismAndroidModule: (main) [0,1780] App context is defined.
[INFO] :   I/System.out: ActivityRecorder: before com.app.z12.Z12Activity@8b9aa05_onCreate(), launchUrl = null
[TRACE] :  Uploaded tiapp metadata with Appcelerator Platform!
[WARN] :   art: Before Android 4.1, method android.graphics.PorterDuffColorFilter android.support.graphics.drawable.VectorDrawableCompat.updateTintFilter(android.graphics.PorterDuffColorFilter, android.content.res.ColorStateList, android.graphics.PorterDuff$Mode) would have incorrectly overridden the package-private method in android.graphics.drawable.Drawable
[INFO] :   I/System.out: ActivityRecorder: after com.app.z12.Z12Activity@8b9aa05_onCreate(), launchUrl = null
[INFO] :   art: Rejecting re-init on previously-failed class java.lang.Class<android.support.v4.app.JobIntentService$JobServiceEngineImpl>
[INFO] :   art: Rejecting re-init on previously-failed class java.lang.Class<android.support.v4.app.JobIntentService$JobServiceEngineImpl>
[DEBUG] :  AndroidRuntime: Shutting down VM
[ERROR] :  TiApplication: (main) [523,2303] No valid root or current activity found for application instance
[INFO] :   art: Starting a blocking GC HeapTrim
```

lokeshchdhry commented:
@ypbnv @garymathews , Tried this on android 8.1.0 as other versions are having some issues. One thing I am seeing here is when I opt out of analytics the `active session` in the `real time` analytics page doe not go to `0`. Is this expected ? 
Also, at the start i.e after the app launch the values of `Ti.Analytics.optedOut` & `Ti.Analytics.getOptedOut()` are `undefined ` & not `false`.

Commit:
[TIMOB-26027] Update aps-analytics.jar

Commit:
[TIMOB-26003] Set Ti.Analytics.optedOut to false

garymathews commented:
@lokeshchdhry Try now 👍

garymathews committed:
Merge branch 'master' into TIMOB-26003

sgtcoolguy committed:
Merge branch 'master' into TIMOB-26003

lokeshchdhry commented:
@garymathews , I am still seeing `undefined` after the app launch.

ypbnv committed:
Merge branch 'master' into TIMOB-26003

ypbnv committed:
Change accessors to optOut property.
Remove BuildConfig from aps_analytics.jar to prevent duplicated depdendcy with ti.cloudush.

ypbnv committed:
Build aps_analytics.jar with the proper language level.

ypbnv commented:
@garymathews , @lokeshchdhry Pushed a version of aps_analytics built with the proper language level.

build commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10058/39/artifact/dist/mobilesdk-7.3.0.v20180529133719-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

garymathews reviewed (approved):
CR: **PASS**

lokeshchdhry committed:
Merge branch 'master' into TIMOB-26003

lokeshchdhry commented:
The `optOut` property seems to work as expected. When `optOut = true` no analytics events are sent from the device & warning is thrown as below.
```
[WARN] :   APSAnalytics: APSAnalytics has not been enabled or the user opted out. Please enable Analytics and ensure the user has not opted out.
```
With `optOut = false`, analytics events are sent successfully.

In depth tests are on this wiki page : https://wiki.appcelerator.org/display/qe/GDPR+Analytics+testing

Studio Ver:     5.1.0.201805300954
SDK Ver:        7.3.0 local build
OS Ver:         10.13.4
Xcode Ver:      Xcode 9.4
Appc NPM:       4.2.13
Appc CLI:       7.0.4-master.2
Daemon Ver:     1.1.3
Ti CLI Ver:     5.1.0
Alloy Ver:      1.12.0
Node Ver:       8.9.1
NPM Ver:        5.5.1
Java Ver:       10
Devices:        ⇨ google Nexus 5 --- Android 6.0.1
                ⇨ google Nexus 6P --- Android 8.1.0
